test/fuzz/transformation_changing_memory_semantics_test.cpp

```diff
@@ -0,0 +1,284 @@
+// Copyright (c) 2021 Mostafa Ashraf
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "source/fuzz/transformation_changing_memory_semantics.h"
+
+#include "gtest/gtest.h"
+#include "source/fuzz/fuzzer_util.h"
+#include "source/fuzz/instruction_descriptor.h"
+#include "test/fuzz/fuzz_test_util.h"
+
+namespace spvtools {
+namespace fuzz {
+namespace {
+TEST(TransformationChangingMemorySemanticsTest, NotApplicable) {
+  const std::string shader = R"(
+               OpCapability Shader
+               OpCapability Int8
+          %1 = OpExtInstImport "GLSL.std.450"
+               OpMemoryModel Logical GLSL450
+               OpEntryPoint Fragment %4 "main"
+               OpExecutionMode %4 OriginUpperLeft
+               OpSource ESSL 320
+          %2 = OpTypeVoid
+          %3 = OpTypeFunction %2
+          %6 = OpTypeInt 32 1
+          %7 = OpTypeInt 8 1
+          %9 = OpTypeInt 32 0
+         %26 = OpTypeFloat 32
+          %8 = OpTypeStruct %6
+         %10 = OpTypePointer StorageBuffer %8
+         %11 = OpVariable %10 StorageBuffer
+         %19 = OpConstant %26 0
+         %18 = OpConstant %9 1
+         %12 = OpConstant %6 0
+         %13 = OpTypePointer StorageBuffer %6
+         %15 = OpConstant %6 4
+         %16 = OpConstant %6 7
+         %17 = OpConstant %7 4
+         %20 = OpConstant %6 16 ; SequentiallyConsistent mode
+         %96 = OpConstant %6 64 ; UniformMemory mode
+         %97 = OpConstant %6 0 ; None mode
+         %98 = OpConstant %6 2 ; Acquire mode
+         %99 = OpConstant %6 4 ; Release mode
+        %100 = OpConstant %6 8 ; AcquireRelease mode
+          %4 = OpFunction %2 None %3
+          %5 = OpLabel
+         %14 = OpAccessChain %13 %11 %12
+         %21 = OpAtomicLoad %6 %14 %15 %20
```

Comment:
I don't think this is quite right. The memory semantics should probably be `SequentiallyConsistent | UniformMemory` otherwise I don't think it has an effect. The same probably applies below. The SPIR-V might currently be valid (or it might be invalid; I am not sure), but regardless, for now, we only really care about _sensible_ uses of memory semantics. Thus, if you are doing an atomic load from a storage buffer, you need to specify `UniformMemory` as part of the memory semantics. This is also helpful for you to understand why your current transformation code is not quite right.

Reply:
Yes @paulthomson, you are right.
I think can made it something like this e.g. `NOT FINAL`
```CPP
  const std::string shader = R"(
               OpCapability Shader
               OpCapability Int8
          %1 = OpExtInstImport "GLSL.std.450"
               OpMemoryModel Logical GLSL450
               OpEntryPoint Fragment %4 "main"
               OpExecutionMode %4 OriginUpperLeft
               OpSource ESSL 320
          %2 = OpTypeVoid
          %3 = OpTypeFunction %2
          %6 = OpTypeInt 32 1
          %7 = OpTypeInt 8 1
          %9 = OpTypeInt 32 0
         %26 = OpTypeFloat 32
          %8 = OpTypeStruct %6
         %10 = OpTypePointer StorageBuffer %8
         %11 = OpVariable %10 StorageBuffer
         %99 = OpTypePointer Workgroup %6
         %100 = OpVariable %99 Workgroup
         %19 = OpConstant %26 0
         %18 = OpConstant %9 1
         %12 = OpConstant %6 0
         %13 = OpTypePointer StorageBuffer %6
         %15 = OpConstant %6 4
         %16 = OpConstant %6 7
         %17 = OpConstant %7 4
         %20 = OpConstant %6 16 ; SequentiallyConsistent mode
         %96 = OpConstant %6 64 ; UniformMemory mode
         %97 = OpConstant %6 0 ; None mode
         %98 = OpConstant %6 2 ; Acquire mode
         %99 = OpConstant %6 4 ; Release mode
        %100 = OpConstant %6 8 ; AcquireRelease mode
          %4 = OpFunction %2 None %3
          %5 = OpLabel
         %14 = OpAccessChain %13 %11 %12
         %21 = OpAtomicLoad %6 %100 %15 %20
         %22 = OpAtomicLoad %6 %100 %15 %98
         %23 = OpAtomicExchange %6 %100 %15 %98 %16
         %25 = OpAtomicCompareExchange %6 %100 %15 %20 %20 %16 %15
         %24 = OpAccessChain %13 %11 %12
               OpAtomicStore %14 %15 %99 %16
               OpReturn
               OpFunctionEnd
  )";
```
It's now a `Workgroup` storage class.
I think it now will be good with memory semantics values.
What do you think @paulthomson?

Reply:
I was considering the first OpAtomicLoad instruction. 

```
%21 = OpAtomicLoad %6 %14 %15 %20
```

The memory semantics operand is `%20`, which is a constant with value `16`.

Looking at section `3.25` in the [SPIR-V specification](https://www.khronos.org/registry/SPIR-V/specs/unified1/SPIRV.pdf) (around page 93), you can see the table of bits for memory semantics. We want the value to be formed from one of: None, Acquire, Release, AcquireRelease, SequentiallyConsistent; **as well as** a value like `UniformMemory`, `WorkgroupMemory`, etc. To combine the values, they should be OR'ed together (binary or), because this value is a bit mask. For example, it looks like you want to use SequentiallyConsistent, and you are accessing a storage buffer.

* SequentiallyConsistent == 0x10 == 16 == `0010000b`
* UniformMemory == 0x40 == 64 == `1000000b`

```
  0010000b
| 1000000b  (binary OR)
  -------- 
  1010000b
```

*   1010000b == 0x50 == 80

So I would expect `%20` to have the value 80.

Reply:
Done.